2011-04-04  Jakub Jelinek  <jakub@redhat.com>

	* toplev.cc (toplev_main_argv): New variable.
	(toplev_main): Initialize it.
	* graphite.cc (init_isl_pointers): Load libisl.so.23 from gcc's private
	directory.

--- gcc/toplev.cc.jj	2008-12-09 23:59:10.000000000 +0100
+++ gcc/toplev.cc	2009-01-27 14:33:52.000000000 +0100
@@ -113,6 +113,8 @@ static void finalize (bool);
 static void crash_signal (int) ATTRIBUTE_NORETURN;
 static void compile_file (void);
 
+const char **toplev_main_argv;
+
 /* Decoded options, and number of such options.  */
 struct cl_decoded_option *save_decoded_options;
 unsigned int save_decoded_options_count;
@@ -2239,6 +2241,8 @@ toplev::main (int argc, char **argv)
 
   expandargv (&argc, &argv);
 
+  toplev_main_argv = CONST_CAST2 (const char **, char **, argv);
+
   /* Initialization of GCC's environment, and diagnostics.  */
   general_init (argv[0], m_init_signals);
 
--- gcc/graphite.cc.jj	2010-12-01 10:24:32.000000000 -0500
+++ gcc/graphite.cc	2010-12-01 11:46:07.832118193 -0500
@@ -64,11 +64,39 @@ __typeof (isl_pointers__) isl_pointers__
 static bool
 init_isl_pointers (void)
 {
-  void *h;
+  void *h = NULL;
+  extern const char **toplev_main_argv;
+  char *buf, *p;
+  size_t len;
 
   if (isl_pointers__.inited)
     return isl_pointers__.h != NULL;
-  h = dlopen ("libisl.so.23", RTLD_LAZY);
+  len = progname - toplev_main_argv[0];
+  buf = XALLOCAVAR (char, len + sizeof "libisl.so.23");
+  memcpy (buf, toplev_main_argv[0], len);
+  strcpy (buf + len, "libisl.so.23");
+  len += sizeof "libisl.so.23";
+  p = strstr (buf, "/libexec/");
+  if (p != NULL)
+    {
+      while (1)
+	{
+	  char *q = strstr (p + 8, "/libexec/");
+	  if (q == NULL)
+	    break;
+	  p = q;
+	}
+      memmove (p + 4, p + 8, len - (p + 8 - buf));
+      h = dlopen (buf, RTLD_LAZY);
+      if (h == NULL)
+	{
+	  len = progname - toplev_main_argv[0];
+	  memcpy (buf, toplev_main_argv[0], len);
+	  strcpy (buf + len, "libisl.so.23");
+	}
+    }
+  if (h == NULL)
+    h = dlopen (buf, RTLD_LAZY);
   isl_pointers__.h = h;
   if (h == NULL)
     return false;
